File: hw_test.m
hw07_worker = hw07();

t0 = 0; tf=3; y0=1;

% test func1
y=hw07_worker.p1(@func1, y0, [t0, tf], 10000, 'euler');
hw_assert( abs (y(end) - exact1(tf)) < 1e-2 );
y=hw07_worker.p1(@func1, y0, [t0, tf], 100, 'midpoint');
hw_assert( abs (y(end) - exact1(tf)) < 1e-2 );
y=hw07_worker.p1(@func1, y0, [t0, tf], 10, 'rk4');
hw_assert( abs (y(end) - exact1(tf)) < 1e-2 );

% test func2
y=hw07_worker.p1(@func2, y0, [t0, tf], 10000, 'euler');
hw_assert( abs (y(end) - exact2(tf)) < 1e-4 );
y=hw07_worker.p1(@func2, y0, [t0, tf], 100, 'midpoint');
hw_assert( abs (y(end) - exact2(tf)) < 1e-4 );
y=hw07_worker.p1(@func2, y0, [t0, tf], 10, 'rk4');
hw_assert( abs (y(end) - exact2(tf)) < 1e-4 );

% test func3
y=hw07_worker.p1(@func3, y0, [t0, tf], 10000, 'euler');
hw_assert( abs (y(end) - exact3(tf)) < 1e-4 );
y=hw07_worker.p1(@func3, y0, [t0, tf], 100, 'midpoint');
hw_assert( abs (y(end) - exact3(tf)) < 1e-4 );
y=hw07_worker.p1(@func3, y0, [t0, tf], 10, 'rk4');
hw_assert( abs (y(end) - exact3(tf)) < 1e-4 );

function ret = func1(t, y) 
    ret = y;
end

function ret = func2(t, y) 
    ret = -y;
end

function ret = func3(t, y) 
    ret = -y + t;
end

function exact = exact1(t)
    exact = exp(t);
end

function exact = exact2(t)
    exact = exp(-t);
end

function exact = exact3(t)
    exact = 2 * exp(-t) + t - 1;
end

function hw_assert(X)
    if X; fprintf('\t PASS\n'); else; fprintf('\t FAIL\n'); end
end

File: hw_test.py
# install pytest first: pip install pytest
# run with the following command: pytest hw_test.py

import numpy as np
from hw07 import *

def test_p1():
    y = p1(func1, 1, [0, 3], 10000, 'euler')
    assert np.isclose(y[-1], exact1(3), atol=1e-2 )
    y = p1(func1, 1, [0, 3], 100, 'midpoint')
    assert np.isclose(y[-1], exact1(3), atol=1e-2 )
    y = p1(func1, 1, [0, 3], 10, 'rk4')
    assert np.isclose(y[-1], exact1(3), atol=1e-2 )
    
def test_p2():
    y = p1(func2, 1, [0, 3], 10000, 'euler')
    assert np.isclose(y[-1], exact2(3), atol=1e-4 )
    y = p1(func2, 1, [0, 3], 100, 'midpoint')
    assert np.isclose(y[-1], exact2(3), atol=1e-4 )
    y = p1(func2, 1, [0, 3], 10, 'rk4')
    assert np.isclose(y[-1], exact2(3), atol=1e-4 )


def test_p3():
    y = p1(func3, 1, [0, 3], 10000, 'euler')
    assert np.isclose(y[-1], exact3(3), atol=1e-4 )
    y = p1(func3, 1, [0, 3], 100, 'midpoint')
    assert np.isclose(y[-1], exact3(3), atol=1e-4 )
    y = p1(func3, 1, [0, 3], 10, 'rk4')
    assert np.isclose(y[-1], exact3(3), atol=1e-4 )
    

def func1(t, y):
    return y

def func2(t, y):
    return -y

def func3(t, y):
    return -y + t


def exact1(t):
    return np.exp(t)

def exact2(t):
    return np.exp(-t)

def exact3(t):
    return 2 * np.exp(-t) + t - 1
